keps/sig-apps/2804-consolidate-workload-controllers-status/README.md

```diff
@@ -0,0 +1,912 @@
+bogged down.
+-->
+
+#### Story 1
+As an end-user of Kubernetes, I'd like all my workload controllers to have consistent states
+
+#### Story 2
+As an developer building Kubernetes Operators, I'd like all my operators deployed to have
+consistent states.
+
+
+### Overview of all conditions
+
+The following table gives an overview on what conditions each of the workload resources support.
+
+|                                    | Progressing | Available |  ReplicaFailure | Suspended | Failed | Complete |
+| ---------------------------------  | ----------- | --------- | --------------- | --------- | ------ | -------- |
+| ReplicaSet & ReplicationController | -           | -         |  failed to create / delete pod (FailedCreate, FailedDelete)  | -         | -           | -        |
+| Deployment                         | True when scaling replicas / creating-updating new ReplicaSet / successfully finished progressing (Pods ready or available for MinReadySeconds). False when failed creating ReplicaSet / reached progressDeadlineSeconds. Unknown when rollout paused | True if if required number of replicas is available (takes MaxSurge and MaxUnavailable into account) | failure propagated from new or old ReplicaSet | -         | -*        | -*          |
+| StatefulSet                        | -           | -         | -               | -         | -      | -        |
+| DaemonSet                          | -           | -         | -               | -         | -      | -        |
+| Job                                | -           | -         | -               | True / False when suspended / resumed | failed execution  (BackoffLimitExceeded, DeadlineExceeded)| completed execution |
+| CronJob**                          | -           | -         | -               | -         | -      | -        |
+
+**\* Success of the rollout is instead represented by a Progressing condition (status and reason)**
+
+**\*\* CronJob does not even have Conditions field in its Status**
+
+### Proposed Conditions
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+We are proposing 4 new conditions to be added to the following controllers:
+- Available (DaemonSet, ReplicaSet & ReplicationController, StatefulSet)
+- Progressing (DaemonSet, StatefulSet)
+- Waiting (Job)
+- Running (Job)
+
+The definitions for Progressing condition (including Failed/Complete) are similar to what we have for [Deployment controller](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#deployment-status).
+
+
+The following table is indicating what conditions are currently available (`X`) and what conditions will be added (`A`).
+
+|                                    | Waiting | Running | Progressing | Available |  ReplicaFailure | Suspended | Failed | Complete |
```

Comment:
I started wondering about squashing `ReplicaFailure` with `Failed` they are somewhat similar

Reply:
Can we deprecate the `ReplicaFailure` condition? Or do you mean we would present the same information in both `Failed` and `ReplicaFailed` conditions?

Reply:
Probably in the long run we could deprecated it, but it'll be quite a while.

Reply:
On a second thought I am not sure we want to go into that. I think it could confuse users since Failed would be relevant only to not that often appearing case when it is not possible to create/delete pods. But users also do care about deployment failure in Progressing condition. I think it might be confusing since Failed condition would not include that, but only the less appearing part (depends on your cluster, but anyway).

I would prefer to omit the squashing from the KEP for now. We can update it if we decide later.

Reply:
I agree with @atiratree suggested, we can update it later.

Reply:
Agreed.